zebra-consensus/src/transaction.rs

```diff
@@ -153,175 +161,243 @@ where
-
-                            async_checks.push(output_rsp.boxed());
-                        }
-
-                        let bvk = sapling_shielded_data.binding_verification_key();
-
-                        // TODO: enable async verification and remove this block - #1939
-                        {
-                            let item: zebra_chain::primitives::redjubjub::batch::Item = (bvk, sapling_shielded_data.binding_sig, &shielded_sighash).into();
-                            item.verify_single().unwrap_or_else(|binding_sig_error| {
-                                let binding_sig_error = binding_sig_error.to_string();
-                                tracing::warn!(%binding_sig_error, "ignoring");
-                                metrics::counter!("zebra.error.sapling.binding",
-                                                  1,
-                                                  "kind" => binding_sig_error);
-                            });
-                            // Ignore errors until binding signatures are fixed
-                            //.map_err(|e| BoxError::from(Box::new(e)))?;
-                        }
-
-                        let _rsp = redjubjub_verifier
-                            .ready_and()
-                            .await?
-                            .call((bvk, sapling_shielded_data.binding_sig, &shielded_sighash).into())
-                            .boxed();
-
-                        // TODO: stop ignoring binding signature errors - #1939
-                        // async_checks.push(rsp);
-                    }
-
-                    // Finally, wait for all asynchronous checks to complete
-                    // successfully, or fail verification if they error.
-                    while let Some(check) = async_checks.next().await {
-                        tracing::trace!(?check, remaining = async_checks.len());
-                        check?;
-                    }
-
-                    Ok(tx.hash())
-                }
-                Transaction::V5 { .. } => {
-                    unimplemented!("v5 transaction validation as specified in ZIP-216, ZIP-224, ZIP-225, and ZIP-244")
+                    Self::verify_v4_transaction(
+                        &req,
+                        network,
+                        script_verifier,
+                        inputs,
+                        joinsplit_data,
+                        sapling_shielded_data,
+                    )
+                    .await
                 }
+                Transaction::V5 { .. } => Self::verify_v5_transaction(&req, network).await,
             }
         }
         .instrument(span)
         .boxed()
     }
 }
+
+impl<ZS> Verifier<ZS>
+where
+    ZS: Service<zs::Request, Response = zs::Response, Error = BoxError> + Send + Clone + 'static,
+    ZS::Future: Send + 'static,
+{
+    async fn verify_v4_transaction(
+        request: &Request,
```

Comment:
Style: `Request` only contains copy types and `Arc`s, so it's cheap to clone if needed:
```suggestion
        request: Request,
```